### PR TITLE
add(feegrant): add feegrant tx and query commands

### PR DESCRIPTION
**Pull request by Zygimantass**
PR still has two issues (not very related to feegrant themselves):
* Can't decode the list of grants (check FIXME)
* Revoke fee grant runs out of gas for some reason

```
go run main.go tx feegrant grant cosmos1lrwqvjeu2eq0wv97d2uka6lzthn7ra6kqxep99 --from test
```

https://www.mintscan.io/cosmos/txs/0414C7F679FC44CDA1EE223E2A318DFBADE0AB24453AAD484F2778A368314638

```
go run main.go q feegrant grant cosmos1lrwqvjeu2eq0wv97d2uka6lzthn7ra6kqxep99
{"granter":"cosmos13a26ypvztxclryjngwj2zhal2kv0eq78zvcdys","grantee":"cosmos1lrwqvjeu2eq0wv97d2uka6lzthn7ra6kqxep99","allowance":{"@type":"/cosmos.feegrant.v1beta1.BasicAllowance","spend_limit":[],"expiration":null}}
````

```
go run main.go q feegrant grants cosmos1lrwqvjeu2eq0wv97d2uka6lzthn7ra6kqxep99
[{"granter":"cosmos13a26ypvztxclryjngwj2zhal2kv0eq78zvcdys","grantee":"cosmos1lrwqvjeu2eq0wv97d2uka6lzthn7ra6kqxep99","allowance":{"@type":"/cosmos.feegrant.v1beta1.BasicAllowance","spend_limit":[],"expiration":null}}]
```

### CODE REVIEW

File: cmd/feegrant.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+			grants, err := cl.QueryFeeGrants(address, pageReq)
+			if err != nil {
+				return err
+			}
+
+			// FIXME: if anyone can tell me why JSON output without this doesn't work gets a 5 OSMO tip
+			// FIXME: seems like there should be no side effects with using the MarshalProto function, but thats not the
+			for grant := range grants {
+				_, err := cl.MarshalProto(grants[grant])
```
Comment (jackzampolin) [inline cmd/feegrant.go:59]: Try to use the marshal interface proto methods on the codec

File: cmd/feegrant.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var (
+				granterAddr sdk.AccAddress
+				granteeAddr sdk.AccAddress
+				err         error
+			)
+
+			cl := config.GetDefaultClient()
+			key, _ := cmd.Flags().GetString(FlagFrom)
```
Comment (jackzampolin) [inline cmd/feegrant.go:130]: Please don't use the FROM flag, lets make these arguements in the style of the other commands.

File: cmd/feegrant.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			if err != nil {
+				return err
+			}
+
+			granteeAddr, err = cl.DecodeBech32AccAddr(args[0])
+			if err != nil {
+				return err
+			}
+
+			spendLimit, err := cmd.Flags().GetString(FlagSpendLimit)
```
Comment (jackzampolin) [inline cmd/feegrant.go:141]: This should be an argument

File: cmd/feegrant.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+			if err != nil {
+				return err
+			}
+
+			spendLimitCoins, err := sdk.ParseCoinsNormalized(spendLimit)
+			if err != nil {
+				return err
+			}
+
+			exp, err := cmd.Flags().GetString(FlagExpiration)
```
Comment (jackzampolin) [inline cmd/feegrant.go:151]: This should also be an arguement

File: cmd/feegrant.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+					return err
+				}
+
+				basicGrant.Expiration = &expiresAtTime
+			}
+
+			var grant feegrant.FeeAllowanceI
+			grant = &basicGrant
+
+			periodClock, err := cmd.Flags().GetInt64(FlagPeriod)
```
Comment (jackzampolin) [inline cmd/feegrant.go:173]: This should be an arguement

File: cmd/feegrant.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+			if err != nil {
+				return err
+			}
+
+			periodLimitVal, err := cmd.Flags().GetString(FlagPeriodLimit)
```
Comment (jackzampolin) [inline cmd/feegrant.go:178]: this should also be an arguement

File: cmd/feegrant.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+					Period:           getPeriod(periodClock),
+					PeriodReset:      getPeriodReset(periodClock),
+					PeriodSpendLimit: periodLimit,
+					PeriodCanSpend:   periodLimit,
+				}
+
+				grant = &periodic
+			}
+
+			allowedMsgs, err := cmd.Flags().GetStringSlice(FlagAllowedMsgs)
```
Comment (jackzampolin) [inline cmd/feegrant.go:213]: Should this be an argument too?

File: cmd/feegrant.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var (
+				granterAddr sdk.AccAddress
+				granteeAddr sdk.AccAddress
+				err         error
+			)
+
+			cl := config.GetDefaultClient()
+			key, _ := cmd.Flags().GetString(FlagFrom)
```
Comment (jackzampolin) [inline cmd/feegrant.go:258]: please don't use the from flag! Lets make the signer an arguement here

File: cmd/feegrant.go (hunk 1 above)
Comment (Zygimantass) [reply cmd/feegrant.go:59]: the interesting thing is that it works with this dummy loop, but it doesn't without it. the loop should theoretically change nothing

File: cmd/feegrant.go (hunk 7 above)
Comment (Zygimantass) [reply cmd/feegrant.go:213]: why should optional args be arguments? seems it'd be complicated to parse them all separately

File: cmd/feegrant.go (hunk 1 above)
Comment (jackzampolin) [reply cmd/feegrant.go]: weird...

**Comment by jackzampolin** [comment]
Hey @Zygimantass I merged my PR into main and deleted the branch and it accidentally closed your PR do you mind reopening against main?

**Pull request closed without merging**